Introduce methods for reading all data across partitions

christianhelle opened the pull request:
I stumble upon a scenario every now and then where I need to re-process ALL data in a given container for things like data migration or for publishing to an EventHub

The changes here introduces the the ReadAllAsync() overload in `ICosmosReader` and `ICosmosBulkReader`

```csharp
/// <summary>
/// Reads all the specified <typeparamref name="T"/> resource from the configured
/// Cosmos collection.
///
/// It is NOT recommended to read data across partitions.
/// This method is provided for convenience for scenarios where ALL data needs to processed or re-processed.
/// </summary>
/// <param name="cancellationToken">The <see cref="CancellationToken"/> used.</param>
/// <returns>An <see cref="IAsyncEnumerable&lt;T&gt;"/> over all the <typeparamref name="T"/> resources.</returns>
public IAsyncEnumerable<T> ReadAllAsync(CancellationToken cancellationToken = default);
```

rickykaare commented:
Great work Chris! 

Originally the idea was to not allow cross-partition queries in the `ICosmosReader`, as these can come with a performance penalty that might not be obvious to the caller. Instead one could use the `IContainerProvider` to access the Cosmos SDK Container directly and "go off-road" to do these kind of things.

However, I do think the support for cross-partition queries would be nice, as these become more frequently used. But I was thinking we could maybe offer them as variants to the `Query` and `PagedQuery` instead of the `ReadAll`? And perhaps with naming that make it very explicit that this will execute in a cross-partition scope.

Something like `CrossPartitionQuery` and `PagedCrossPartitionQuery`.

christianhelle commented:
> Great work Chris!
> 
> Originally the idea was to not allow cross-partition queries in the `ICosmosReader`, as these can come with a performance penalty that might not be obvious to the caller. Instead one could use the `IContainerProvider` to access the Cosmos SDK Container directly and "go off-road" to do these kind of things.
> 
> However, I do think the support for cross-partition queries would be nice, as these become more frequently used. But I was thinking we could maybe offer them as variants to the `Query` and `PagedQuery` instead of the `ReadAll`? And perhaps with naming that make it very explicit that this will execute in a cross-partition scope.
> 
> Something like `CrossPartitionQuery` and `PagedCrossPartitionQuery`.

@rickykaare I can do that

christianhelle commented:
@rickykaare I refactored a bit and introduced `CrossPartitionQueryAsync()` and `CrossPartitionPagedQueryAsync()` in ICosmosReader and ICosmosBulkReader

rickykaare reviewed (approved):
Looks good :-) Great work Chris!
